src/storage/ccloudPreloader.ts

```diff
@@ -81,50 +74,56 @@ export class CCloudResourcePreloader {
-        // queue up to fetch all the schemas plus mainly tickle the side effect of
-        // storing those schemas into the resource manager.
-        promises.push(getSchemas(envGroup.environment, schemaRegistry.id));
+    // Start loading the ccloud-related resources from sidecar API into the resource manager for local caching.
+    // If the loading fails at any time (including, say, the user logs out of CCloud while in progress), then
+    // an exception will be thrown and the loadingComplete flag will remain false.
+    try {
+      const resourceManager = getResourceManager();
+
+      // Fetch the from-sidecar-API list of triplets of (environment, kafkaClusters, schemaRegistry)
+      const envGroups = await getEnvironments();
+
+      // Queue up to store the environments in the resource manager
+      const environments = envGroups.map((envGroup) => envGroup.environment);
+      await resourceManager.setCCloudEnvironments(environments);
+
+      // Collect all of the Kafka clusters into a single array and queue up to store them in the resource manager.
+      // (Each environment may have many clusters.)
+      const kafkaClusters = envGroups.flatMap((envGroup) => envGroup.kafkaClusters);
+      await resourceManager.setCCloudKafkaClusters(kafkaClusters);
+
+      // Likewise the schema registries, but filter out any undefineds for environments that don't have one.
+      const schemaRegistries: SchemaRegistryCluster[] = envGroups
+        .map((envGroup) => envGroup.schemaRegistry)
+        .filter(
+          (schemaRegistry): schemaRegistry is SchemaRegistryCluster => schemaRegistry !== undefined,
+        );
+
+      await resourceManager.setCCloudSchemaRegistryClusters(schemaRegistries);
+
+      // For each environment, if there's a schema registry, queue up fetching (and caching) its schemas.
+      // Is safe to fetch + cache each environment's schemas in parallel.
+      const promises: Promise<Schema[]>[] = [];
+
+      for (const envGroup of envGroups) {
+        const schemaRegistry = envGroup.schemaRegistry;
+        if (schemaRegistry !== undefined) {
+          // queue up to fetch all the schemas plus mainly tickle the side effect of
+          // storing those schemas into the resource manager.
+          promises.push(getSchemas(envGroup.environment, schemaRegistry.id));
+        }
       }
-    }
-    await Promise.all(promises);
+      await Promise.all(promises);
 
-    // TODO: add flink compute pools here?
+      // TODO: add flink compute pools here?
 
-    // All done, clear the promise and mark the loading as complete.
-    this.currentlyLoadingPromise = null;
-
-    // In case the user logged out _while_ we were busy loading, err on setting loadingComplete to false.
-    this.loadingComplete = this._hasCCloudConnection;
+      // If made it to this point, all the resources have been fetched and cached and can be trusted.
+      this.loadingComplete = true;
+    } finally {
```

Comment:
Should we add a basic `catch` and some logging in here just in case?

Reply:
Sure.